### PR TITLE
[BACKLOG-6733] Initial implementation of the configuration service

### PR DESCRIPTION
**Pull request by nantunes**
@dcleao please take a look.

### CODE REVIEW

**Comment by brosander** [comment]
http://wingman.pentaho.com:8080/job/wingman/34/

**Comment by brosander** [comment]
http://wingman.pentaho.com:8080/job/wingman/37/

**Comment by brosander** [comment]
Base build failed with return code 1

File: package-res/resources/web/pentaho/type/config/ConfigurationService.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      if (config.rules) {
+        config.rules.forEach(function(rule) {
+          this.addRule(rule);
+        }, this);
+      }
+    },
+
+    addRule: function(rule) {
+      var select = rule.select || {};
+      var typeIds = select.type || ["pentaho/type/Instance"];
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/config/ConfigurationService.js]: "pentaho/type/value"

File: package-res/resources/web/pentaho/type/config/ConfigurationService.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+define([
+  "../../lang/Base"
+], function(Base) {
+  "use strict";
+
+  var ConfigurationService = Base.extend("pentaho.type.config.ConfigurationService", /** @lends pentaho.type.config.ConfigurationService#*/{
+    /**
+     * @private
+     */
+    ruleStore: {},
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/config/ConfigurationService.js]: I would preferably use the "_", for safety.

File: package-res/resources/web/pentaho/type/config/ConfigurationService.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        this.ruleStore[typeId].push(rule);
+      }, this);
+    },
+
+    select: function(typeId, context) {
+      // TODO Select the apropriate rules, merge them and return
+
+      // Temporary placeholder mock implementation
+      // always return first configuration (or empty, if none)
+      var rules = [];
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/config/ConfigurationService.js]: `var configs` or configSpecs ?

File: package-res/resources/web/pentaho/type/config/AmdLoadedConfigurationService.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  "./ConfigurationService",
+  "pentaho/service!pentaho.type.ITypeConfiguration?meta"
+], function(ConfigurationService, configurations) {
+  "use strict";
+
+  configurations.sort(function(a, b) {
+    return a.moduleId.localeCompare(b.moduleId);
+  });
+
+  var AmdLoadedConfigurationService = ConfigurationService.extend("pentaho.type.config.AmdLoadedConfigurationService", /** @lends pentaho.type.config.AmdLoadedConfigurationService#*/{
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/config/AmdLoadedConfigurationService.js]: This `@lends` implies we'll be adding some doclets?

Comment (dcleao) [inline package-res/resources/web/pentaho/type/config/AmdLoadedConfigurationService.js]: Not sure about the name, but, honestly, can't think of a better one.

File: test-js/unit/pentaho/type/configurationService.Spec.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Must be executed out of strict scope
+var __global__ = this;
```
Comment (dcleao) [inline test-js/unit/pentaho/type/configurationService.Spec.js]: Lost variable...

File: test-js/unit/pentaho/type/configurationService.Spec.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        localRequire.define("testModule2", "testModuleValue2");
+        localRequire.define("testModule3", "testModuleValue3");
+      }, function(configurationService) {
+        expect(["testModuleValue1", "testModuleValue2", "testModuleValue3"]).toContain(configurationService);
+      });
+    });
+
+    it("returns by default a instance of ConfigurationService when no service is configured", function() {
+
+      return require.using(["pentaho/type/configurationService"], function(configurationService) {
```
Comment (dcleao) [inline test-js/unit/pentaho/type/configurationService.Spec.js]: Don't think this test can pass as the top-level ConfigurationService is a class loaded in a different RequireJS context. No?

**Comment by dcleao** [comment]
👍

File: test-js/unit/pentaho/type/configurationService.Spec.js (hunk 6 above)
Comment (nantunes) [inline test-js/unit/pentaho/type/configurationService.Spec.js]: Yep... Noticed that before reading this, so some minutes wasted. ;-)

**Comment by nantunes** [comment]
Rebased. Depends on #728 to be merged.

@pentaho/millenniumfalcon please take a moment to address this.

**Pull request merged**